docs/CppInteroperability/SwiftToCppInteroperabilityVisionDocument.md

```diff
@@ -0,0 +1,137 @@
+
+This document is a prospective feature vision document, as described in the [draft review management guidelines](https://github.com/rjmccall/swift-evolution/blob/057b2383102f34c3d0f5b257f82bba0f5b94683d/review_management.md#future-directions-and-roadmaps) of the Swift evolution process. It has not yet been approved by the Language Workgroup.
+
+## Introduction
+
+Swift currently provides support for interoperability with C and Objective-C. Swift can import C and Objective-C APIs, and it can also expose its `@objc` types and functions to Objective-C. However, Swift currently does not provide support for interoperability with C++. Supporting bidirectional C++ interoperability in Swift would allow it to call into C++ APIs, and would make Swift APIs accessible to C++. Making Swift APIs accessible to C++ would simplify the process of adoption of Swift in existing C++ codebases, as it would allow them to incrementally adopt Swift. Furthermore, it would make Swift-only libraries available to C++ codebases.
+
+## Overview of Swift's interoperability support
+
+Swift uses two very different approaches for interoperability with C and Objective-C. For "forward" interoperability, Swift embeds a copy of the Clang compiler, which it uses to directly load C and Objective-C Clang modules and translate declarations into a native Swift representation. For "reverse" interoperability, Swift does not want to require Clang to embed Swift, and so Swift generates an Objective-C header that can be used to call into Swift APIs that are exposed in that header. Because Objective-C is restricted in what kinds of types it can work with, and because it requires code to be emitted in a special way that doesn't match the regular Swift ABI, methods and types must be marked as being exposed to Objective-C with the `@objc` attribute.
+
+Swift's support for C++ interoperability is modeled after its support for C and Objective-C interoperability. For "forward" interoperability, the embedded Clang compiler is used to directly load C++ Clang modules and translate declarations into a native Swift representation. The "forward" interoperability [vision document](https://github.com/apple/swift/pull/60501/files) provides more details about this model. For "reverse" interoperability, Swift generates a header that uses C++ language constructs to represent Swift APIs that are exposed by the Swift module. Because C++ is much more expressive and flexible than Objective-C, the generated header is able to provide representation for native Swift functions, methods, initializers, accessors and types. This allows C++ programmers to call into Swift APIs using the familiar C++ function and member function call syntax.
+
+### C++ interoperability is opt-out not opt-in
+
```

Comment:
The Language Workgroup feels that it would be valuable to center the philosophy behind this decision.  It's abstractly always better for interop to work the way that you're proposing for C++: the foreign language consumes the public interface of the Swift code and makes whatever it can available without needing extra code to be emitted by the Swift compiler.  Objective-C interop is specifically constrained by the need to (1) use ObjC-compatible classes and (2) put Objective-C methods directly in their method tables, which inherently requires extra code (and that code has to be in the Swift module unless you generate an extra level of ObjC object wrapper for everything, which would violate other goals).  Those constraints don't apply to C++, so we do the more abstractly preferable thing.

I think you can rework the following paragraph to lead into that philosophy after the first sentence.  You can then talk about things like adding support for the Swift calling convention as enabling a C++ compiler to directly interoperate with a Swift module without extra code.

Reply:
I rewrote this section following your comment.